charms/katib-db-manager/src/charm.py

```diff
@@ -136,17 +158,132 @@ def _check_leader(self):
+                    "Failed to create relational-db relation due to existing mysql relation"
+                )
+                return
+        except KeyError:
+            # mysql relation was not found, proceed
+            pass
+        self._on_event(event)
+
+    def _on_relational_db_relation_removed(self, event):
+        """Remove relational-db relation data."""
+        pass
+
+    def _get_db_data(self) -> dict:
+        """Check for MySQL relations -  mysql or relational-db - and retrieve data.
+        
+        Only one database relation can be established at a time.
+        """
+        db_data = {}
+        relation_data = {}
+        try:
+            # retrieve mysql relation data
+            relation = self.model.get_relation("mysql")
+            if not relation:
+                # myslq relation is not established
+                # raise an error and proceed to check for relational-db relation
+                # this error is processed below
+                raise GenericCharmRuntimeError("Database relation mysql is not established")
             unit = next(iter(relation.units))
-            mysql_data = relation.data[unit]
-            # Ensure we've got some data sent over the relation
-            mysql_data["root_password"]
-        except (IndexError, StopIteration, KeyError):
-            raise ErrorWithStatus("Waiting for mysql connection information", WaitingStatus)
-
-        return mysql_data
+            relation_data = relation.data[unit]
+            # retrieve database data from relation data
+            # this also validates the expected data by means of KeyError exception
+            db_data["db_name"] = self._database_name
+            db_data["db_username"] = relation_data["user"] # need to verify
+            db_data["db_password"] = relation_data["root_password"]
+            db_data["katib_db_host"] = relation_data["host"]
+            db_data["katib_db_port"] = relation_data["port"]
+            db_data["katib_db_name"] = relation_data["database"]
+        except (IndexError, StopIteration, KeyError) as err:
+            # mysql relation is established
+            # failed to retrieve database configuration
+            if len(relation_data) == 0:
+                self.logger.info("Found empty relation data for mysql relation.")
+                raise ErrorWithStatus("Waiting for mysql data", WaitingStatus)
+            self.logger.error(
+                f"Missing attribute {err} in mysql relation data: {relation_data}"
+            )
+            # incorrect/incomplete data can be found in mysql relation which can be resolved:
+            # use WaitingStatus
+            raise ErrorWithStatus(
+                f"Incorrect/incomplete data found in relation mysql. See logs", WaitingStatus
+            )
+        except GenericCharmRuntimeError:
+            # at this point it is detected that mysql relation is not established
+            # check for relational-db relation
```

Comment:
I think we need two different functions. One will handle mariadb, another mysql. That will help us in future when you will remove deprecated interface. Then we just delete one method. But now it is a spaghetti monster with mixing of all the stuff together :D